akka-persistence-typed/src/main/scala/akka/persistence/typed/dstate/scaladsl/Effect.scala

```diff
@@ -0,0 +1,112 @@
+/*
+ * Copyright (C) 2021 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.persistence.typed.dstate.scaladsl
+
+import akka.actor.typed.ActorRef
+import akka.annotation.DoNotInherit
+import akka.persistence.typed.dstate.internal
+import akka.persistence.typed.internal.SideEffect
+
+/**
+ * A command handler returns an `Effect` directive that defines the next State to persist.
+ *
+ * Instances are created through the factories in the [[Effect]] companion object.
+ *
+ * Not for user extension.
+ */
+@DoNotInherit
+trait Effect[+State]
+
+trait ReplyEffect[+State] extends Effect[State]
+
+trait EffectBuilder[+State] extends Effect[State] {
+
+  /**
+   * Run the given callback. Callbacks are run sequentially.
+   */
+  def thenRun(callback: State => Unit): EffectBuilder[State]
+
+  /** The side effect is to stop the actor */
+  def thenStop(): EffectBuilder[State]
+
+  /**
+   * Unstash the commands that were stashed with [[Effect.stash]].
+   *
+   * It's allowed to stash messages while unstashing. Those newly added
+   * commands will not be processed by this `unstashAll` effect and have to be unstashed
+   * by another `unstashAll`.
+   */
+  def thenUnstashAll(): Effect[State]
+
+  def thenReply[ReplyMessage](replyTo: ActorRef[ReplyMessage])(
+      replyWithMessage: State => ReplyMessage): ReplyEffect[State]
+
+  def thenNoReply(): ReplyEffect[State]
+}
+
+/**
+ * Factory methods for creating [[Effect]] directives - how an durable state actor reacts on a command.
+ */
+object Effect {
+
+  /**
+   * Insert or update.
+   */
+  def persist[State](state: State): EffectBuilder[State] =
+    internal.Persist(state)
+
+  /**
+   * Major difference with respect to EventSourced, a durable state can be deleted
+   *
+   * A delete must trigger a stop or reload the empty state
+   * In any case, it needs to start from scratch.
+   */
```

Comment:
I think it would be nice if this always triggered stop and creating a new empty slate for same id would have to be starting a new actor instance for that id.

One tricky aspect, just like with event sourced, might be deleted and not existing anymore vs tombstone and remembered that it was deleted. Not sure if we need API for that though, good to mention in the docs when we get there.

Reply:
I think we should go for a hard delete with the possibility to re-create it. If the user wants a tombstone, then they should switch to a tombstone state and stop accepting new commands. And yes, we should mention that this in the docs.

Reply:
Explicit tombstone being user responsibility sounds good.

I'm still thinking that it could make sense if hard delete should always mean stop actor, since that would automatically make starting it again be a new state/clean slate and avoid passing some state between deleted and new actor (especially for Java which will be a class based API to follow the lead of the event sourced API)

I may remember wrong, but for `EventSourcedBehavior` I think we don't have a delete-entire-entity API but rather defer such needs to journal specific APIs. In the classic API we have `deleteTo` but that is more of an imperative clean up events after snapshots, which we made a higher level abstraction for in typed. If I think about how such an API for `EventSourcedBehavior` would look think it would make sense there as well to have completion make the actor stop.